### DISCUSSION TITLE
SSH Key Session Control

### DISCUSSION THREAD

**Discussion by PowershellScripter** [kingToolbox/WindTerm#153] [Q&A]
Is there a way to bind an ssh key to a session? I have ssh key auth turned on for my linux box and created keys on my windows system, but everytime I connect, I have to manually go to the location and choose the private key to use. Is there a ways to create a session to auto connect with the private key location path?

**Comment by kingToolbox** [comment]
Are you using the latest version? The latest [WindTerm_2.1.0_Windows_Portable](https://github.com/kingToolbox/WindTerm/releases/tag/2.1) version only needs to fill in the ssh key path once and it should automatically remember the key path.

**Comment by PowershellScripter** [comment]
@kingToolbox 
Sorry for the late response, never got notified of your reply. I am on the 2.1.0 portable and I go to log in and enter my user and the next screen says "Agent authentication failed, please try another method".  I choose my identity file and it logs me in, then when I exit it asks me if I want to save the session. I choose yes, then re-open it and go to log back into that session as same user, but it A. Still asks me to enter a user so it doesnt remember that part and B. Does not remember the ssh file path last used as it still fails and I have to manually choose the file again.

**Reply by kingToolbox** [reply]
First of all, if you check the `auto login next time`, then as long as you successfully log in once, the subsequent logins will automatically log in. Secondly, even if you do not check the `auto login next time`, for `public-key authentication`, it is the same. As long as you successfully log in once, the subsequent logins will automatically log in too. WindTerm will remember the path of the identity file. Either way, the prerequisite is that you must successfully log in once.

I tested multiple logins using `public-key authentication` and everything is ok. So if you can't log in automatically, please try to check the `auto login next time`, it should solve the problem. 

In addition, `Agent authentication failed, please try another method` indicates that `Pageant` is not started, so `agent authentication` cannot be performed. And the question of whether to save sessions when exiting refers to whether it is necessary to restore all current sessions at the next startup. Both are not related to `public-key authentication`.

**Reply by PowershellScripter** [reply]
@kingToolbox 
I verified the host is known (green) and checked the auto login next time check box and put in my user account and clicked continue, next page says `Agent authentication failed, please try another method` so I chose the identity file for the private key, then hit continue `Successfully logged in`. From here, I closed the console (tried both saving and not saving session). Opened the console, still the same issue. I can do an OBS screen recording if youd like.

**Reply by kingToolbox** [reply]
This is strange. Please follow the steps below to check the configuration file. 

- Use the identity file to successfully log in. 
- Close WindTerm.
- Open `profiles/default.v10/terminal/user.sessions`.
- Find your session and confirm whether there is `ssh.identityFilePtah`, and its value is the path of your identity file. 

Please make sure that the path consists entirely of English letters. Although this is not necessary, just in case.

**Reply by PowershellScripter** [reply]
@kingToolbox 
No such entry exists in my user.sessions file at that path. I tried entering the sessions with the ssh key again and nothing. I even tried renaming the file to trigger what I thought would be a new sessions file creation and it did not create a new file or modify the renamed file. The only file that changed was the reopen.history file.

**Reply by PowershellScripter** [reply]
@kingToolbox 
I just downloaded a fresh copy of WindTerm and repeated the process you requested and same result, no `ssh.identityfilepath` in the `user.sessions` file

**Reply by kingToolbox** [reply]
- Close WindTerm.
- Manually add `"ssh.identityFilePath" : "c:\\ssh_path\\your_id_rsa"`  to your session.
- Start WindTerm. 

Please try to add the configuration manually using the steps above to see if it works? 

Sorry, I haven't responded to you in time in the past few days.  I am still working hard on development. I hope to complete the development and release the WindTerm_2.2.0_Prerelase_1 version as soon as possible. There are many internal changes in this version, so I think it may solve your problem. Even if it doesn't, it will make more sense to solve it directly in the new version.

**Reply by PowershellScripter** [reply]
@kingToolbox
Okay, so I literally just figured out what the issue was this whole time.
If I just type an address into the connectionbar and connect that way, I can add password, ssh key, etc but once I connect, it doesnt log any of that information into user.sessions. But, if I go to file > new session, and create an ssh session that route, it WILL save all the info in user.sessions. Now I know how to go about this in the future, but im curious, is this by design? or is it supposed to save this info in user.sessions even when connecting through the connection bar? If it is, its not doing that. But if its not, ive been using WindTerm wrong the whole time lol.

By the way, you may not remember me, but I was the person defending you on reddit months and months ago when people were trying to badger you about the security implementations of your application being closed sourced. 
Not telling you this to get any special treatment or bragging or anything like that, just letting you know who I am so you know im a long time user of your product and it is by far the best putty replacement Ive ever used. 
I recommend your application to everyone who uses ssh, telnet, etc.
Sad part of using it for so long is that its taken me this long to figure out how to use it properly when it comes to getting user sessions saved lol.
I think the confusing part of it was that it would ask me if I wanted it to remember the session information for the next time I would log in, which I as the end user would assume it would automatically make sure it saved all that info automatically and re-initialized it on start up. But it doesnt do that if you create the connection from the connection bar.

Is that possibly something that CAN be done or is in development?
